Static support

marcdel opened the pull request:
Hey @toddw / @cybrass,

I wasn't sure whether to PR against `develop` or `toddw/static-support`, but the later was a bit noisy since the branch is out of date. Happy to make another one there instead if you'd rather though.

The feature itself is now working (I was able to create/build a [test app](http://gluestick-static-test.s3-website-us-east-1.amazonaws.com/) with my branch and upload the `build` folder to S3 and it works great), and normal start/build commands seem to be working again as well.

I could use some guidance on the tests. I'm a bit stumped writing the tests for generateStaticHTMLFile (see the PR [here](https://github.com/marcdel/gluestick/pull/1) for weirdo webpack issues). Also, some of the tests I wrote for RequestHandler.prepareOutput may not be necessary and there may be some that I missed. Open to any suggestions.

toddw committed:
beginning support for static builds

marcdel committed:
Pass staticBuild to prepareOutput in RequestHandler tests.

marcdel committed:
Test that the body is empty when staticBuild is true.

marcdel committed:
Write the response output to index.html in the build folder.

Commit:
Use the values from assets param to set the script/link sources.

Commit:
Don't require Entry on module load.

Commit:
Use entry point rather than assets in getHead.

Commit:
Remove unused assets param from getHead method.

Commit:
Remove webPackIsomorphicTools from prepareOutput since it's no longer used.

Commit:
Early load Entry in production only. Breaks elsewhere.

Commit:
Test that isEmail and current state are not set when staticBuild true.

I added these for completeness, but I'm not sure how much value they
really bring. I'm not sure we care whether they're set or not. I think
the if statement on line 153 could be removed without any negative
consequences (maybe a slight performance hit, but it doesn't really matter
since we're only going to be running prepareOutput once).

marcdel commented on src/lib/server/RequestHandler.js:
```diff
@@ -27,7 +26,7 @@ let _Entry;
  * is initialized but instead when the method is run.
  */
-if (process.env.NODE_ENV !== "test") {
+if (process.env.NODE_ENV === "production") {
```
This is potentially controversial since I'm not sure the original reason it was added. As far as I can tell now that `generateStaticHTMLFile` is being imported in build.js this will fail unless NODE_ENV is production anyway.

Commit:
Remove unused variables to make `npm run lint` happy.

toddw committed:
Merge branch 'develop' into static-support

Commit:
#392 add config to aliases

Commit:
script to update package version

Commit:
#378 exclude test from start commant

ferrannp committed:
#406: Include Jest and starting moving tests to it (removing expect and sinon).

sethbattin committed:
force posix-form for template's __$PATH__ property

christinebrass committed:
Merge branch 'develop' into static-support

Commit:
#392 add config to aliases

zamotany committed:
#276: provide CSS vendor budle in production

Commit:
Added expressConfigurator function to expose express app

zamotany committed:
#276: unsure process.env is always defined

ferrannp committed:
#406: Second round of converting tests to use Jest expect and mocks.

zamotany committed:
#322: remove node-inspector from dependencies

zamotany committed:
#322: use built in inspector when debugging server

zamotany committed:
#276: remove redundant code

everything work fine without it

michalchudziak committed:
Webpack config hook.

michalchudziak committed:
Remove comment, and add missing package.json

michalchudziak committed:
Set source-map as default production devtool

ferrannp committed:
#406: All tests are just now depending on Jest (no sinon no expect).

ferrannp committed:
#406: Remove old tests scripts from package.json.

ferrannp committed:
Merge remote-tracking branch 'origin/develop' into 406_refactor_test_suite

Conflicts:
	package.json

michalchudziak committed:
Fix test.

ferrannp committed:
#406: Fix linting issues in test dir and add Jest into .eslintrc config as env + remove unnecessary globals.

zamotany committed:
#322: --debug-brk must be always passed to node

debugger statements are triggered only when inspector is connected

toddw committed:
Merge branch 'develop' into 276_fix_bundle_generation

toddw committed:
Merge pull request #430 from callstack-io/276_fix_bundle_generation

Fix bundle generation

toddw committed:
Merge pull request #436 from callstack-io/355_expose_express_app

Added expressConfigurator function to expose express app

toddw committed:
Merge pull request #433 from callstack-io/407_webpack_config_exposal

Webpack config exposal

toddw committed:
Merge pull request #432 from callstack-io/392_config_to_aliases

392 config to aliases

toddw committed:
Merge pull request #426 from TrueCar/sbattin/windows-generate

force posix-form for template's __$PATH__ property

toddw committed:
Merge pull request #435 from callstack-io/140_version_update_script

Script to update package version

toddw committed:
Merge pull request #431 from callstack-io/378_exclude_test_from_start

#378 exclude test from start commant

zamotany committed:
#322: add jest env config

zamotany committed:
#322: remove node-inspector mentions

zamotany committed:
Merge branch '406_refactor_test_suite'

Conflicts:
	templates/new/package.json

zamotany committed:
#322: debug gluestick test command + readme

zamotany committed:
#322: debug application test with gluestick cli

ferrannp committed:
Merge branch 'develop' into 406_refactor_test_suite

Conflicts:
	package.json

ferrannp committed:
#406: Mock prompt for new.test.js which was causing Jest to be hanged up.

zamotany committed:
#419: custom tag function for template

zamotany committed:
#419: linting fixes

zamotany committed:
Merged branch 406_refactor_test_suite into 322_built_in_inspector

michalchudziak committed:
[WIP] Move to webpack 2.2

zamotany committed:
#419: config parser implementation

zamotany committed:
#419: (example-like) predefined component generator

ferrannp committed:
#442: Add success message for "new".

michalchudziak committed:
Working webpack 2.2 config.

zamotany committed:
#419: integrate with CLI

zamotany committed:
#419: write template to file

[WIP] only implementation without test

zamotany committed:
#419: component generator fix

toddw committed:
support `v1` ,`v2` type branches in travis

toddw committed:
Merge pull request #437 from callstack-io/406_refactor_test_suite

Refactor test suite to Jest (+watch +coverage)

zamotany committed:
Merged branch v1 into 322_built_in_inspector

toddw committed:
remove debug-brk from server debugging, watch for changes

toddw committed:
Merge pull request #445 from callstack-io/322_built_in_inspector

Use built-in  node inspector instead of node-inspector

christinebrass committed:
Merge pull request #453 from TrueCar/toddw/updates-to-inspector

remove debug-brk from server debugging, watch for changes

michalchudziak committed:
Merge.

ferrannp committed:
Merge branch 'v1' into 442-success-message-2

Conflicts:
	test/commands/new.test.js

michalchudziak committed:
Make webpack production builds work.

michalchudziak committed:
Dir patch quickfix.

michalchudziak committed:
Update template to use new webpack.

michalchudziak committed:
Merge branch 'v1' of github.com:TrueCar/gluestick into mike866/webpack_upgrade

ferrannp committed:
#442: Add "gluestick start" + rewrite test to Jest.

Commit:
#410 onEnter hook added

michalchudziak committed:
Remove preloaders hook.

Commit:
#409 single entrypoint

zamotany committed:
#419: make camelCase and kebab-case interop

zamotany committed:
#419: update predefined component generator

zamotany committed:
#419: generator for custom generator

ferrannp committed:
#427: Solving getWebpackAdditions test.

zamotany committed:
#419: update generator for generator

zamotany committed:
#419: initial implmementation of reducer generator

zamotany committed:
#419: improved API

zamotany committed:
#419: fix component generator

zamotany committed:
#419: add reducer generator

zamotany committed:
#419: update generator for generator to match new API

zamotany committed:
#419: container generator + component generator fix

zamotany committed:
#419: remove old templates

zamotany committed:
#419: missing test suites

zamotany committed:
Merged branch v1 into zamotany/419_generator_api

toddw committed:
Merge pull request #450 from TrueCar/442-success-message-2

`gluestick new` success message

toddw committed:
fix `config` alias

Alias were being defined in two locations. We couldn't consolodate them
into webpack-shared-config because it imported
webpack-isomorphic-tools-config, so I consoldated them there instead.

Now there is only 1 place to update alias

toddw committed:
Merge pull request #459 from TrueCar/toddw/fix-alias

fix `config` alias

toddw committed:
add httpClient support on server side

toddw committed:
Merge pull request #455 from TrueCar/410_onEnter_hook

#410 onEnter hook added

christinebrass committed:
Merge branch 'develop' into static-support

zamotany committed:
#419: update reducer generator

zamotany committed:
#419: replace modificator to modifier

zamotany committed:
#419: replace spec with test in generator test files

zamotany committed:
#419: use webpack aliases

zamotany committed:
#419: remove first newline in template + update tests

zamotany committed:
#419: handler additional directory from CLI

zamotany committed:
#419: output written and modified files path

zamotany committed:
#419: rewrite names in component and reducer generator

zamotany committed:
#419: add check for alphanumberical name

zamotany committed:
#419: fix failing assertion

zamotany committed:
Merged branch v1 into 409_support_single_entrypoint

Commit:
upgraded dependencies

Commit:
Fixed eslint errors

ferrannp committed:
#447: Generated app is using and running Jest for testing.

zamotany committed:
entrypoints CLI option

zamotany committed:
Merge branch '409_backup'

Conflicts:
	src/cli.js
	src/commands/start-client.js

zamotany committed:
#409: update tests

ferrannp committed:
#444: Fix destroy test (async/await issue).

ferrannp committed:
Merge pull request #464 from TrueCar/lukewalczak/fix_eslint

Fixed eslint errors

ferrannp committed:
Merge branch 'v1' into lukewalczak/deps_upgrade

ferrannp committed:
#447: Add --coverage, --watch and --pattern options.

zamotany committed:
#419: reorder generator path to allow to override predefined gens

Commit:
upgraded templates package.json

michalchudziak committed:
Bin command.

michalchudziak committed:
Expose bin from node modules to CLI command.

ferrannp committed:
#447: Debug Jest tests in generated app.

michalchudziak committed:
Quickfix.

toddw committed:
Merge pull request #458 from TrueCar/zamotany/419_generator_api

Generator API + custom generators

toddw committed:
Merge pull request #471 from TrueCar/mike866/expose-bin

Expose node_modules/.bin/ directory trough `gluestick bin` command.

Commit:
#409 update entrypoints

Commit:
#409 fix tests and mapEntryToGroup

Commit:
#409 refactor

Commit:
Merge branch 'v1' into 409_support_single_entrypoint

Commit:
Added options to pass to the generator

ferrannp committed:
#447: Remove all unused old testing related code and dependencies.

toddw committed:
Merge pull request #456 from TrueCar/mike866/webpack_upgrade

Webpack Upgrade.

toddw committed:
Merge pull request #472 from TrueCar/lukewalczak/469

Added options to pass to the generator

ferrannp committed:
Merge branch 'v1' into ferrannp/generated-app-jest

Conflicts:
	src/config/karma-config.js

ferrannp committed:
#447: Remove karma-config as part of merge conflict.

ferrannp committed:
Merge pull request #465 from TrueCar/ferrannp/generated-app-jest

Move generated app to use Jest as test solution

Commit:
Merge branch 'v1' into lukewalczak/deps_upgrade

Commit:
Merge branch 'v1' into 409_support_single_entrypoint

michalchudziak committed:
Merge pull request #463 from TrueCar/lukewalczak/deps_upgrade

Upgrade dependencies

Commit:
fixed useless dependencies

ferrannp committed:
Merge pull request #486 from TrueCar/lukewalczak/deps_fix

Fixed old dependencies that were re-added after merge

marcdel committed:
Merge branch 'develop' into static-support

zamotany committed:
Merged branch v1 into 409_support_single_entrypoint

zamotany committed:
#409: fix failing vendors bundle

zamotany committed:
#409: resolve path if needed

zamotany committed:
#409: test and linting fixes

toddw committed:
add src/index.js to addon regex

Commit:
Remove extra loadjs var

toddw committed:
Merge pull request #512 from TrueCar/v1-cherry-pick

V1 cherry pick

toddw committed:
Merge pull request #457 from TrueCar/409_support_single_entrypoint

#409 single entrypoint

toddw committed:
bump version

toddw committed:
bump version

toddw committed:
yarn updates

ferrannp committed:
#518: Fix broken generate command and add test for it.

ferrannp committed:
#515: Update generator test template to use Jest.

toddw committed:
Merge pull request #520 from TrueCar/ferrannp/v1-generate-templates-test

[V1] Update generator test template to use Jest

toddw committed:
Merge pull request #519 from TrueCar/ferrannp/v1_broken_generate_command

[V1] Broken generate command

Commit:
Fix loadjs typo

toddw committed:
[BUG] fix expressConfigurator

The original PR that adds the expressConfigurator added it to
start-client, but start-client only runs in development mode.

This PR simply moves it to where it belongs.

Original PR: https://github.com/TrueCar/gluestick/pull/436/files

christinebrass committed:
Merge pull request #524 from TrueCar/toddw/fix-express-configurator

[BUG] fix expressConfigurator

christinebrass committed:
Merge branch 'develop' into static-support

christinebrass commented:
Looks like the latest build of this contains issues with the `-Z` option. Somehow `config.server` is `undefined` in this case only, causing this line to throw: https://github.com/marcdel/gluestick/blob/0a7356e70a3e6a75cb4c5773cf5e19ae96ae8bb3/src/lib/server/getHead.js#L37

marcdel committed:
Pass default config values when building with -Z and --static.

Fixes undefined config.server.loadjs in getHead.js:37:48

christinebrass commented:
Looks good! We'll just want to base this PR on top of `v1` now and test any other changes that might be included. 👍

marcdel committed:
Merge remote-tracking branch 'upstream/v1' into static-support

marcdel committed:
Add webpack back as an argument and update tests for jest compat.

marcdel committed:
Fixes warnings for RequestContext required props.

marcdel committed:
Fix warnings about Body required props in tests.

marcdel committed:
Fixes issue in which entry points for path undefined with --static flag

For some reason when you run `NODE_ENV=production gluestick build
--static` getWebpackEntries returns an object in a different
configuration than when you run `NODE_ENV=production gluestick start`.

`NODE_ENV=production gluestick build --static`
{ entryPoints: { mapEntryToGroup: {}, entries: { '/': [Object] } },
path: '/' }
entryPoints.entries[path] { name: 'main',
  routes: '/Users/marc/Code/marcdel/gs-static-test/src/config/routes',
  reducers: '/Users/marc/Code/marcdel/gs-static-test/src/reducers',
  fileName: 'main',
  filePath:
    '/Users/marc/Code/marcdel/gs-static-test/src/config/.entries/main-[chunkhash].js',
  index: '/Users/marc/Code/marcdel/gs-static-test/Index',
  entryPoints: null }

`gluestick start`

{ entryPoints:
  { '/':
    { name: 'main',
routes:
      '/Users/marc/Code/marcdel/gs-static-test/src/config/routes',
      reducers:
        '/Users/marc/Code/marcdel/gs-static-test/src/reducers',
      fileName: 'main',
      filePath:
        '/Users/marc/Code/marcdel/gs-static-test/src/config/.entries/main-[chunkhash].js',
      index:
        '/Users/marc/Code/marcdel/gs-static-test/Index',
      entryPoints:
        null } },
  path: '/' }
  entryPoints[path]
{ name:
  'main',
    routes:
      '/Users/marc/Code/marcdel/gs-static-test/src/config/routes',
    reducers:
      '/Users/marc/Code/marcdel/gs-static-test/src/reducers',
    fileName:
      'main',
    filePath:
      '/Users/marc/Code/marcdel/gs-static-test/src/config/.entries/main-[chunkhash].js',
    index:
      '/Users/marc/Code/marcdel/gs-static-test/Index',
    entryPoints:
      null
}

marcdel committed:
'entryPoint' is never reassigned. Use 'const' instead.

marcdel commented:
Closing in favor of https://github.com/TrueCar/gluestick/pull/539
